### PR TITLE
Update the ParticleNet section

### PR DESCRIPTION
**Pull request by colizz**
A WIP PR to update the ParticleNet section.

Todo list:
 - [x] update the training subsection (make sure the code and explanations are consistent)
 - [x] update the full architecture plot
 - [x] [`weaver-benchmark`](https://github.com/colizz/weaver-benchmark): fix the file importing bug
 - [x] [`weaver-benchmark`](https://github.com/colizz/weaver-benchmark): rewrite the script for top dataset conversion file
 - [x] redo the LR finder test with recommended settings and update the section
 - [x] add tips in multiple class training
 - [x] push new features to `Weaver` (integrating Tensorboard)

### CODE REVIEW

**Comment by colizz** [comment]
(I fixed the condor script a little. Now the section should be ready)
Thanks for reviewing!

File: content/inference/particlenet.md [hunk 1]
```diff
@@ -86,35 +86,48 @@ Recent works in the joint field of HEP and ML also shed light on exploiting the
...
+Our goal is to do some warm-up with `Weaver`, and more importantly, to explore from a technical side the neural net architectures, from a simple multi-layer perceptron (MLP) model, to a more complicated "DeepAK8 tagger" model based on 1D CNN with ResNet, and eventually to the "ParticleNet model" which is based on DGCNN. We will dig deeper into their implementations in `Weaver` and try to illustrate as many details as possible. Finally, we compare their performance and see if we can reproduce the benchmark record with the model. Please clone the repo `weaver-benchmark` and we'll get started. The `Weaver` repo will be cloned as a submodule.
```
Comment (aperloff) [inline content/inference/particlenet.md:89]: Our goal is to do some warm-up with `Weaver`, and more importantly, to explore from a technical side the neural net architectures: a simple multi-layer perceptron (MLP) model, a more complicated "DeepAK8 tagger" model based on 1D CNN with ResNet, and the "ParticleNet model," which is based on DGCNN.

File: content/inference/particlenet.md [hunk 2]
```diff
@@ -86,35 +86,48 @@ Recent works in the joint field of HEP and ML also shed light on exploiting the
...
+
+# Create a soft link inside weaver so that it can find data/model cards
+ln -s ../top_tagging weaver-benchmark/weaver/top_tagging
+```
 
 
 ### 1. Build models in `Weaver`
 
-When implementing a new training in `Weaver`, two key elements are crucial: the model and the data configuration file. The model configuration file includes a `get_model` function that returns a `torch.nn.Module` type model and a dictionary of model info used to export an ONNX-format model. The data configuration is a YAML file describing how to process the input data. Please see the `Weaver` README for details.
+When implementing a new training in `Weaver`, two key elements are crucial: the model and the data configuration file. The model defines the network architecture we are using, and the data configuration includes which variables to use for training, which pre-selection to apply and how to assign truth labels, etc.
```
Comment (aperloff) [inline content/inference/particlenet.md:100]: "...apply, how to assign truth labels, etc."

File: content/inference/particlenet.md [hunk 3]
```diff
@@ -86,35 +86,48 @@ Recent works in the joint field of HEP and ML also shed light on exploiting the
...
+Before moving on, we need a preprocessing of the benchmark datasets. The original sample is an H5 file including branches like energy `E_i` and 3-momenta `PX_i`, `PY_i`, `PZ_i` for each jet constituent *i* (*i*=0, ..., 199) inside a jet. All branches are in the 1D flat structure. We reconstruct the data in a way that the jet features are 2D vectors (e.g., in the `vector<float>` format): `Part_E`, `Part_PX`, `Part_PY`, `Part_PZ`, with variable-length that corresponds to the number of constituents. Note that this is a commonly used data structure, similar to the NanoAOD format in CMS.
 
-The input files after preprocessing (in the `.awkd` format) can be found at CERN EOS space `/eos/user/c/coli/public/weaver-benchmark/samples/top/`. It includes three sets of data for training, validation, and test.
+The datasets can be found at CERN EOS space `/eos/user/c/coli/public/weaver-benchmark/top_tagging/samples`. The input files used in this page are in fact the ROOT files produced by the preprocessing step, stored under `prep/` subdirectory. It includes three sets of data for training, validation, and test.
```
Comment (aperloff) [inline content/inference/particlenet.md:106]: "...under the `prep/` subdirectory."

File: content/inference/particlenet.md [hunk 4]
```diff
@@ -179,7 +192,7 @@ Then, we show three NN model configurations below and provide detailed explanati
             return torch.nn.CrossEntropyLoss()
         ```      
 
-    Below show the full structure of the MLP network printed by PyTorch. You will see it in the `Weaver` output during the training.
+    Below shows the full structure of the MLP network printed by PyTorch. You will see it in the `Weaver` output during the training.
```
Comment (aperloff) [inline content/inference/particlenet.md:195]: "The output below shows..."

File: content/inference/particlenet.md [hunk 5]
```diff
@@ -357,7 +451,7 @@ Then, we show three NN model configurations below and provide detailed explanati
             return torch.nn.CrossEntropyLoss()
         ```
     
-    Below show the full structure of the DeepAK8 model based on 1D CNN with ResNet. It is printed by PyTorch and you will see it in the `Weaver` output during training.
+    Below shows the full structure of the DeepAK8 model based on 1D CNN with ResNet. It is printed by PyTorch and you will see it in the `Weaver` output during training.
```
Comment (aperloff) [inline content/inference/particlenet.md:454]: "The output below shows..."

File: content/inference/particlenet.md [hunk 6]
```diff
@@ -993,76 +1095,77 @@ Then, we show three NN model configurations below and provide detailed explanati
...
+     --network-config top_tagging/networks/${MODEL_CONFIG} \
+     --model-prefix output/${PREFIX} \
+     --gpus 0,1 --batch-size 1024 --start-lr 5e-3 --num-epochs 20 --optimizer ranger \
+     --log output/${PREFIX}.train.log
     ```
 
-    Here are several caveats running this command.
+    Here `--gpus 0,1` specifies the GPUs to run with the device ID 1 and 2. For training on CPUs, please use `--gpu ''`.
+    
+    A detailed description of the training command can be found in [`Weaver` README](https://github.com/hqucms/weaver). We would say a few more words in the data loading options, because it depends on the scenarios we feed in the input data. Here are several caveats.
```
Comment (aperloff) [inline content/inference/particlenet.md:1138]: Below we will note a few more caveats about the data loading options, though the specific settings will depend on the specifics of the input data.

File: content/inference/particlenet.md [hunk 7]
```diff
@@ -993,76 +1095,77 @@ Then, we show three NN model configurations below and provide detailed explanati
...
+
+    !!! warning "Caveats on the data loading options"
+        Our goal in data loading is to guarantee that the data loaded in every mini-batch is evenly distributed with different labels, though they are not necessarily stored evenly in the file. Besides, we also need to ensure that the on-the-fly loading and preprocessing of data should be smooth and not be a bottleneck of the data delivering pipeline. The total amount of loaded data also need to be controlled so as not to explode the entire memory. These all guide us on how we choose the correct option.
```
Comment (aperloff) [inline content/inference/particlenet.md:1141]: "also need to be" -> "also needs to be"

Comment (aperloff) [reply content/inference/particlenet.md:1141]: Instead of "These all guide us on how we choose the correct option," try, "The following guidelines should be used to choose the best options for your use case:"

File: content/inference/particlenet.md [hunk 8]
```diff
@@ -1088,69 +1195,76 @@ Here we present three ways of training. For readers who have a local machine wit
...
+        Modify the argument line. These are the bash variable `PREFIX`, `MODEL_CONFIG`, `DATA_CONFIG`, `PATH_TO_SAMPLES` used in the `Weaver` command. Since the EOS directory is accessable accross all condor nodes on lxplus, one may directly specify `<your-path-to-samples>` as the EOS path provided above. An example is shown in the commented line.
+        ```bash linenums="1" hl_lines="3"
         Universe                = vanilla
         executable              = run.sh
-        arguments               = 
-        output                  = logs/$(ClusterId).$(ProcId).out
-        error                   = logs/$(ClusterId).$(ProcId).err
-        log                     = logs/$(ClusterId).log
+        arguments               = <prefix> <model-config> <data-config> <your-path-to-samples>
+        #arguments              = mlp mlp_pf.py pf_features.yaml /eos/user/c/coli/public/weaver-benchmark/top_tagging/samples
```
Comment (aperloff) [inline content/inference/particlenet.md:1242]: Why do you have a line here which is commented out?

File: content/inference/particlenet.md [hunk 9]
```diff
@@ -1258,6 +1381,20 @@ The Tensorboard monitoring plots in the training/evaluation loss is shown as fol
...
+
+The preselection should be chosen in a way that all remaining events passing the selection should fall into one and only one category. In other words, events with no labels attached should not be kept since it will confuse the training process.
+
+Class weights (the `class_weights` option under `weights` in the data config) control the relative importance of input sample categories for training. Implementation-wise, it changes the event possibility in a specific category chosen as training input events. The class weight comes into effect when one trains a multi-class classifier. Take 3-class case (denoted as [A, B, C]) as an example, the `class_weights: [1, 1, 1]` gives equal weights to all categories. Retraining the input with `class_weights: [10, 1, 1]` may result in a better discriminating power for class A vs. B or A vs. C; while the power of B separating with C will be weakened. As a trade-off between separating A vs. C and B vs. C, the class weights need to be intentionally tuned to achieve reasonable performance.
```
Comment (aperloff) [inline content/inference/particlenet.md:1392]: Did you mean to use the word "possibility," or did you mean "probability"?

File: content/inference/particlenet.md [hunk 10]
```diff
@@ -1258,6 +1381,20 @@ The Tensorboard monitoring plots in the training/evaluation loss is shown as fol
...
+
+Despite the different options for the class weights, one method to factor out the interplay across categories is to define "binarized" score between two classes. Suppose the raw score for the three classes are *P*(A), *P*(B), and *P*(C) (their sum should be 1), then one can define the discriminant *P*(BvsC) = *P*(B) / (*P*(B)+*P*(C)) to separate B vs. C. In this way, the saparating power of B vs. C will remain unchanged for `class_weights` configured as either `[1, 1, 1]` or `[10, 1, 1]`. This strategy has been widely used in CMS to define composite tagger discrimant which are applied analysis-wise.
```
Comment (aperloff) [inline content/inference/particlenet.md:1394]: Did you mean to use the word "despite," which indicated that the next section can be done whether or not class weights are used? Or did you mean something like "Instead of using class weights, one method..."? That would indicate that this is a second option, instead of using class weights.

File: content/inference/particlenet.md (hunk 8 above)
Comment (colizz) [reply content/inference/particlenet.md:1242]: Here we are trying to provide a real configuration example, so people will not doubt whether to add quote marks etc in a condor configuration line. This is mentioned above: "An example is shown in the commented line."

File: content/inference/particlenet.md (hunk 10 above)
Comment (colizz) [reply content/inference/particlenet.md]: Previously I thought it was "despite", but a better strategy is to apply both options. The sentence is changed

**Comment by aperloff** [comment]
@colizz Is this PR ready to be merged?

**Comment by colizz** [comment]
All suggestions are applied. Please take a further look :)

**Comment by colizz** [comment]
> @colizz Is this PR ready to be merged?

yes, it is ready to merge
